### PR TITLE
chore(frontend): distribute index by arrange key

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

## What's changed and what's your intention?

As title

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

close https://github.com/singularity-data/risingwave/issues/2017

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs [hunk 1]
```diff
@@ -99,17 +99,25 @@ impl StreamMaterialize {
...
         user_order_by: Order,
         user_cols: FixedBitSet,
+        distribute_only_order_by: bool,
     ) -> Result<Self> {
         // ensure the same pk will not shuffle to different node
         let input = match input.distribution() {
             Distribution::Single => input,
-            _ => Distribution::HashShard(input.pk_indices().to_vec())
-                .enforce_if_not_satisfies(input, Order::any()),
+            _ => Distribution::HashShard(if distribute_only_order_by {
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/plan_node/stream_materialize.rs:115]: Looks somehow strange to me... I think this should be handled by the caller, not here.

File: e2e_test/v2/streaming/index.slt [hunk 2]
```diff
@@ -35,12 +35,10 @@ select v4, v1, v3, v5 from iii_index;
 4 3 3 3
 5 2 2 2
 
-halt
-
 # TODO: support drop index
 
 statement ok
-drop index iii_index_t1;
+drop materialized view iii_index_t1;
```
Comment (fuyufjh) [inline e2e_test/v2/streaming/index.slt:41]: May let `drop index` act as an alias of `drop materialized view` in the future

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs (hunk 1 above)
Comment (skyzh) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:115]: Infer distribution is always done inside operators?

Comment (fuyufjh) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:115]: Hmmm. Also makes sense

**Review by fuyufjh — APPROVED** [review]

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs (hunk 1 above)
Comment (skyzh) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:115]: In fact I'm also considering making `Materialize` to be `AnyShard`, and let the create mv / index statement to decide the distribution by specifying it in PlanRoot. But in this case, we might need some redesign.

**Pull request merged**